ujt/callbacks.py

```diff
@@ -608,102 +717,117 @@ def save_tag(n_clicks_timestamp, input_values):
-        add_timestamps: List of the timestamps of the create-tag-button buttons was called.
+        delete_tag_signal: Signal indicating that a tag was deleted.
             Value unused, input only provided to register callback.
-
+        
     Returns:
-        A signal to add to the tag-update-signal hidden div.
+        A list of components to be placed in the tag-panel.
     """
+    return components.get_create_tag_components()
 
-    ctx = dash.callback_context
-    triggered_id, triggered_prop, triggered_value = utils.ctx_triggered_info(ctx)
-
-    # When the button is initially added, it fires a callback.
-    # We want to prevent this callback from making changes to the update signal.
-    if triggered_value is None:
-        raise PreventUpdate
-
-    # Unfortunately, we have to convert the stringified dict back to a dict.
-    # Dash doesn't provide us any other method to see which element triggered the callback.
-    # This isn't very elegant, but I don't see any other way to proceed.
-    id_dict = utils.string_to_dict(triggered_id)
-    if "create-tag-button" in id_dict:
-        state.create_tag("")
-    elif "delete-tag-button" in id_dict:
-        tag_idx = id_dict["index"]
-        state.delete_tag(tag_idx)
-    else:
-        raise ValueError
 
-    return constants.OK_SIGNAL
+#endregion
 
 
+#region apply tag panel
 @app.callback(
-    Output("tag-panel",
+    Output("add-applied-tag-signal",
            "children"),
-    Input("tag-update-signal",
-          "children"),
+    Input({"add-applied-tag-button": ALL},
+          "n_clicks_timestamp"),
+    [
+        State("cytoscape-graph",
+              "tapNode"),
+        State("cytoscape-graph",
+              "tapEdge"),
+    ],
+    prevent_initial_call=True,
 )
-def generate_tag_panel(tag_update_signal):
-    """ Handles generating the tag creation and deletion panel.
+def apply_new_empty_tag(add_timestamps, tap_node, tap_edge):
+    """ Handles applying a new empty tag to the tag map.
 
     This function is called:
-        when a new tag is created.
-        when a tag is deleted.
+        when the add applied tag button is clicked.
 
     Args:
-        tag-update-signal: Signal indicating that the list of tags was modified.
+        add_timestamps: List of the timestamps of the add-applied-tag-button buttons was called.
             Value unused, input only provided to register callback.
-        
+            Should only contain one value.
+        tap_node: The cytoscape element of the latest tapped node.
+        tap_edge: The cytoscape element of the latest tapped edge.
+
     Returns:
-        A list of components to be placed in the tag-panel.
+        A signal to add to the add-applied-tag-signal hidden div.
     """
-    return components.get_tag_panel()
+
+    ctx = dash.callback_context
+    triggered_id, triggered_prop, triggered_value = utils.ctx_triggered_info(ctx)
+
+    # When the button is initially added, it fires a callback.
```

Comment:
These four lines are repeated a lot, does it make sense to extract out a method called preventOnAddUpdate or something so it's just one repeated line?

Reply:
The first two:
```
ctx = dash.callback_context
triggered_id, triggered_prop, triggered_value = utils.ctx_triggered_info(ctx)
```
are necessary whenever we have callbacks that need to check the source of the callback. I think these should be combined into one line (skip the assignment of ctx) -- will do that all at once in a future branch.

We could combine the last two lines 
```
if triggered_value is None:
        raise PreventUpdate
```
into prevent_on_add_update(triggered_value) which could make the semantics clearer without needing the comment there each time. 

However I don't think we should combine all four because there's many cases where we need triggered_id, and the other properties but still want to update the output value on the initial callback.